src/role_towerdrainer.js

```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+/**
+ * Towerdrainer is used to drain energy from hostile towers
+ *
+ * Moves to the border between routing targetRoom and attackRoom,
+ * step for one tick into attackRoom, then step out ang heals
+ *
+ * You need 13 towerdrainers to drain at maximum speed
+ *
+ * Towerdrainers doesn't called now from anywhere. Call tehm manually:
+ * @example
+ * Game.rooms.E17N1.memory.queue.push({role: 'towerdrainer', routing: {targetRoom: 'E16N0'}, attackRoom: 'E16N1'})
+ */
+
+roles.towerdrainer = {};
+
+roles.towerdrainer.settings = {
+  layoutString: 'TMH',
+  amount: [3, 5, 2], // attack RCL 5
+  // amount: [2, 3, 1], // attack RCL 3
+  maxLayoutAmount: 1,
+};
+
+roles.towerdrainer.action = function(creep) {
+  const restRoom = creep.memory.routing.targetRoom;
+  const attackRoom = creep.memory.attackRoom;
+
+  if (!creep.memory.restPosition) {
+    creep.notifyWhenAttacked(false);
+    const room = Game.rooms[restRoom];
+    const attackDirrection = room.findExitTo(attackRoom);
+    const restDirrection = (attackDirrection + 3) % 8 + 1;
+    const occupiedPositions = {};
+    _.filter(Game.creeps, c => c.memory.role === 'towerdrainer' && c.memory.restPosition).forEach(c => {
+      occupiedPositions[c.memory.restPosition.x + c.memory.restPosition.y] = c.id;
+    });
+    const attackExits = room.find(attackDirrection);
+    for (let exit of attackExits) {
+      const pos = exit.getAdjacentPosition(restDirrection);
+      if (!pos.checkForWall() && !pos.checkForObstacleStructure() && !occupiedPositions[pos.x + pos.y]) {
+        creep.memory.restPosition = pos;
+        creep.memory.attackDirrection = attackDirrection;
+        creep.memory.restDirrection = restDirrection;
+        break;
+      }
+    }
+  }
+
+  const restPos = creep.memory.restPosition;
+  if (!restPos) {
+    creep.log('no position');
+    creep.moveRandom();
+    return false;
+  }
+
+  if (creep.pos.roomName === attackRoom || creep.pos.isBorder(-1) && creep.pos.isNearTo(restPos.x, restPos.y)) {
+    creep.heal(creep);
+    creep.move(creep.memory.restDirrection);
+  } else if (creep.hits < creep.hitsMax) {
+    creep.heal(creep);
+  } else if (creep.pos.isEqualTo(restPos.x, restPos.y)) {
+    creep.move(creep.memory.attackDirrection);
+  } else {
+    creep.moveTo(restPos.x, restPos.y);
+  }
+
+  return true;
+};
+
+roles.towerdrainer.execute = function(creep) {
```

Comment:
`.execute` is the old system. I think this can be executed within the creep execution layer. I'm usually adding it to see when it is executed. Just as information.